events/collector.go

```diff
@@ -0,0 +1,255 @@
+package events
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/mongodb/ftdc"
+	"github.com/pkg/errors"
+)
+
+// Collector wraps the ftdc.Collector interface and adds
+// specific awareness of the Performance type from this package. These
+// collectors should be responsible for cumulative summing of values,
+// when appropriate.
+//
+// In general, implementations should obstruct calls to underlying
+// collectors Add() method to avoid confusion, either by panicing or
+// by no-oping.
+type Collector interface {
+	AddEvent(*Performance) error
+	ftdc.Collector
+}
+
+type basicCumulativeCollector struct {
+	ftdc.Collector
+	current *Performance
+}
+
+// NewBasicCollector produces a collector implementation that adds
+// Performance points to the underlying FTDC collector. Counter values
+// in the point are added to the previous point so that the values are
+// cumulative.
+//
+// This event Collector implementation captures the maximal amount of
+// fidelity and should be used except when dictated by retention
+// strategy.
+func NewBasicCollector(fc ftdc.Collector) Collector {
+	return &basicCumulativeCollector{
+		Collector: fc,
+	}
+}
+
+func (c *basicCumulativeCollector) Add(interface{}) error { return nil }
+func (c *basicCumulativeCollector) AddEvent(in *Performance) error {
+	if in == nil {
+		return errors.New("cannot add nil performance event")
+	}
+	if c.current == nil {
+		c.current = in
+		return c.Collector.Add(c.current.MarshalDocument())
+	}
+
+	c.current.Add(in)
+	return c.Collector.Add(c.current.MarshalDocument())
+}
+
+type passthroughCollector struct {
+	ftdc.Collector
+}
+
+// NewPassthroughCollector constructs a collector that does not sum
+// Performance events and just passes them directly to the underlying
+// collector.
+func NewPassthroughCollector(fc ftdc.Collector) Collector {
+	return &passthroughCollector{
+		Collector: fc,
+	}
+}
+
+func (c *passthroughCollector) Add(interface{}) error { return nil }
+func (c *passthroughCollector) AddEvent(in *Performance) error {
+	if in == nil {
+		return errors.New("cannot add nil performance event")
+	}
+
+	return c.Collector.Add(in.MarshalDocument())
+}
+
+type samplingCollector struct {
+	ftdc.Collector
+	current *Performance
+	sample  int
+	count   int
+}
+
+// NewSamplingCollector constructs a collector that has the same
+// semantics as the basic, adding all sampled documents together, but
+// only persisting every n-th sample to the underlying collector.
+func NewSamplingCollector(fc ftdc.Collector, n int) Collector {
+	return &samplingCollector{
+		sample:    n,
+		Collector: fc,
+	}
+}
+
+func (c *samplingCollector) Add(interface{}) error { return nil }
+func (c *samplingCollector) AddEvent(in *Performance) error {
+	if in == nil {
+		return errors.New("cannot add nil performance event")
+	}
+
+	if c.current == nil {
+		c.current = in
+		c.count++
+		return c.Collector.Add(c.current.MarshalDocument())
```

Comment:
Why are we adding to the collector here? Shouldn't we check `shouldCollect` first? Or we always want to add the first performance event?

Reply:
because presumably this is the first point, and we have to have something saved in the collector to add/cumulative to.